Add cert-manager documentation (#1317)
minio/docs: source/operations/cert-manager/cert-manager-operator.rst

.. _minio-certmanager-operator:

=========================
cert-manager for Operator
=========================

.. default-domain:: minio

.. contents:: Table of Contents
   :local:
   :depth: 1

> MinIO Operator manages the TLS certificate issuing for the services hosted in the ``minio-operator`` namespace. 

Reviewer: maybe
```suggestion
MinIO Operator manages TLS certificate issuing for the services hosted in the ``minio-operator`` namespace. 
```

> This page describes how to manage the Operator's TLS certificates with :ref:`cert-manager <minio-certmanager>`.

Reviewer 2: This is more

```suggestion
This procedure assumes a new or fresh installation of the MinIO Operator configured to use cert-manager for TLS certificate management.
```

We should also consider how to document migrating an existing Operator config to cert-manager.

Author: Asking about existing installs internally. We can address that in a separate PR.
I modified your suggestion a bit, but did add that this is assuming a new/fresh install of Operator.

Author: Added a section on migrating an existing deployment.

Prerequisites
-------------

- A `supported version of Kubernetes <https://kubernetes.io/releases/>`__. 
- `kustomize <https://kustomize.io/>`__ installed
- ``kubectl`` access to your ``k8s`` cluster
- Completed the steps to :ref:`set up cert-manager <minio-setup-certmanager>`
- The MinIO Operator must not yet be installed.

1) Create a CA Issuer for the ``minio-operator`` namespace
----------------------------------------------------------

This guide **disables** the automatic generation of certificates in MinIO Operator and issues certificates using cert-manager instead.

> The ``minio-operator`` namespace needs to have its own certificate authority (CA), derived from the cluster's ``ClusterIssuer`` certificate created during :ref:`cert-manager setup <minio-certmanager>`.

Reviewer: ```suggestion
The ``minio-operator`` namespace must have its own certificate authority (CA), derived from the cluster's ``ClusterIssuer`` certificate created during :ref:`cert-manager setup <minio-certmanager>`.
```

Create this CA certificate using cert-manager.

.. important::

>    This CA certificate should exist *before* installing MinIO Operator.

Reviewer: maybe this could be `must` also, it won't work otherwise I think?
```suggestion
   This CA certificate must exist *before* installing MinIO Operator.
```

1. If it does not exist, create the ``minio-operator`` namespace

   .. code-block:: shell
      :class: copyable

      kubectl create ns minio-operator

2. Request a new Certificate with ``spec.isCA: true`` specified. 

   This certificate serves as the :abbr:`CA (Certificate Authority)` for the `minio-operator` namespace.

   Create a file called ``operator-ca-tls-secret.yaml`` with the following contents:

   .. code-block:: yaml
      :class: copyable
      :emphasize-lines: 7,8

      # operator-ca-tls-secret.yaml
      apiVersion: cert-manager.io/v1
      kind: Certificate
      metadata:
        name: minio-operator-ca-certificate
        namespace: minio-operator
      spec:
        isCA: true
        commonName: operator
        secretName: operator-ca-tls
        duration: 70128h # 8y
        privateKey:
          algorithm: ECDSA
          size: 256

>         issuerRef:

Reviewer 2: This must match the name of the ClusterIssuer created in the previous guide right?

We should state that after the YAML, along with a sentence to the effect of 'If you specified a different ClusterIssuer name *or* are using a different Issuer from the guide, modify the `issuerRef` to match your environment.'

Author: Ah, yes, good catch!

2) Use the secret to create the ``Issuer``
------------------------------------------

Use the secret created above to add an ``Issuer`` resource for the ``minio-operator`` namespace.

1. Create a file called ``operator-ca-issuer.yaml`` with the following contents: 

   .. code-block:: yaml

      # operator-ca-issuer.yaml
      apiVersion: cert-manager.io/v1
      kind: Issuer
      metadata:
        name: minio-operator-ca-issuer
        namespace: minio-operator
      spec:
        ca:
          secretName: operator-ca-tls

2. Apply the resource to the cluster
   
   .. code-block:: shell

      kubectl apply -f operator-ca-issuer.yaml

3) Create TLS certificate
-------------------------

Now that the ``Issuer`` exists in the ``minio-operator`` namespace, cert-manager can add a certificate.

The certificate from cert-manager must be valid for the following DNS domains:

- ``sts``
- ``sts.minio-operator.svc.``
- ``sts.minio-operator.svc.<cluster domain>``

  .. important::

      Replace ``<cluster domain>`` with the actual value for your MinIO tenant.
      ``cluster domain`` is the internal root DNS domain assigned in your Kubernetes cluster. 

>       Typically, this is ``cluster.local``, but confirm the value by checking your coredns configuration for the correct value for your Kubernetes cluster. 

Reviewer: Is coredns here referring to a process or the CoreDNS product?

Author: The product, I think.

      Check with your Kubernetes provider for more information.

1. Create a ``Certificate`` for the domains mentioned above:

   Create a file named ``sts-tls-certificate.yaml`` with the following contents:

   .. code-block:: yaml
      :class: copyable
      :emphasize-lines: 7,12

      # sts-tls-certificate.yaml
      apiVersion: cert-manager.io/v1
      kind: Certificate
      metadata:
        name: sts-certmanager-cert
        namespace: minio-operator
      spec:
        dnsNames:
          - sts
          - sts.minio-operator.svc
          - sts.minio-operator.svc.cluster.local # Replace cluster.local with the value for your domain.
        secretName: sts-tls
        issuerRef:
          name: minio-operator-ca-issuer

   .. important::
   
      The ``spec.secretName`` is not optional.
   
      The secret name **must** be ``sts-tls``.
      Confirm this by setting ``spec.secretName: sts-tls`` as highlighted above.

2. Apply the resource to the cluster:

   .. code-block:: shell
      :class: copyable

      kubectl apply -f sts-tls-certificate.yaml

This creates a secret called ``sts-tls`` in the ``minio-operator`` namespace.

.. warning::
  

>    Failing to provide the ``sts-tls`` secret containing the TLS certificate or providing an invalid ``key-value`` pair in the secret will prevent the STS service from starting.

Reviewer: A possible version in case this warning should be more alarming
```suggestion
   The STS service will not start if the ``sts-tls`` secret, containing the TLS certificate, is missing or contains an invalid ``key-value`` pair. 
```

Author: I'll take that suggestion.

4) Install Operator with Auto TLS disabled
------------------------------------------

You can now :ref:`install the MinIO Operator <minio-operator-installation>`.

When installing the Operator deployment, set the ``OPERATOR_STS_AUTO_TLS_ENABLED`` environment variable to ``off`` in the ``minio-operator`` container. 

Disabling this environment variable prevents the MinIO Operator from issuing the certificates.

> Instead, Operator waits for cert-manager to issue the TLS certificate.

Reviewer 2: Interesting - do we always wait for cert-manager, no other configs required?

Author: Swapping "waits for" for "relies on".

> Instead, Operator waits for the TLS certificate issued by cert-manager.

Reviewer: maybe
```suggestion
Instead, Operator waits for cert-manager to issue the TLS certificate.
```

> There are several for defining an environment variable depending on how you install the Operator.

Reviewer: Maybe this was intended to say "several methods" ? Not entirely sure.

Author: Something like that. I went with "various methods".

      metadata:
        name: minio-operator-ca-certificate
        namespace: minio-operator
      spec:
        isCA: true
        commonName: operator
        secretName: operator-ca-tls
        duration: 70128h # 8y
        privateKey:
          algorithm: ECDSA
          size: 256
        issuerRef:
          name: selfsigned-root
          kind: ClusterIssuer
          group: cert-manager.io

   .. important::

      The ``spec.issueRef.name`` must match the name of the ``ClusterIssuer`` created when :ref:`setting up cert-manager <minio-cert-manager-create-cluster-issuer>`.
      If you specified a different ``ClusterIssuer`` name or are using a different ``Issuer`` from the guide, modify the ``issuerRef`` to match your environment.

3. Apply the resource:
   
   .. code-block:: shell
      :class: copyable

      kubectl apply -f operator-ca-tls-secret.yaml

Kubernetes creates a new secret with the name ``operator-ca-tls`` in the ``minio-operator`` namespace.

.. important::

   Make sure to trust this certificate in any applications that need to interact with the MinIO Operator.

2) Use the secret to create the ``Issuer``
------------------------------------------

> Use the secret created above to add an ``Issuer`` resource for the ``minio-operator`` namespace.

Reviewer 3: ```suggestion
Use the secret you just created to add an ``Issuer`` resource for the ``minio-operator`` namespace.
```
Not sure you need "you just created", but working with the confirmation of step approach you've already got going.

Author: Used the secret's actual name and removed the "created above" text. I like it better. Thanks!

Reviewer 3: Even better 🙏

        namespace: minio-operator
      spec:
        ca:
          secretName: operator-ca-tls

2. Apply the resource:
   
   .. code-block:: shell

      kubectl apply -f operator-ca-issuer.yaml

3) Create TLS certificate
-------------------------

Now that the ``Issuer`` exists in the ``minio-operator`` namespace, cert-manager can add a certificate.

The certificate from cert-manager must be valid for the following DNS domains:

- ``sts``
- ``sts.minio-operator.svc.``
- ``sts.minio-operator.svc.<cluster domain>``

  .. important::

      Replace ``<cluster domain>`` with the actual value for your MinIO tenant.
      ``cluster domain`` is the internal root DNS domain assigned in your Kubernetes cluster. 
      Typically, this is ``cluster.local``, but confirm the value by checking your CoreDNS configuration for the correct value for your Kubernetes cluster. 
      
      For example:

      .. code-block:: shell
         :class: copyable

         kubectl get configmap coredns -n kube-system -o jsonpath="{.data}"

      Different Kubernetes providers manage the root domain differently.
      Check with your Kubernetes provider for more information.

> 1. Create a ``Certificate`` for the domains mentioned above:

Reviewer 3: ```suggestion
1. Create a ``Certificate`` for the specified domains:
```
directional text

   Create a file named ``sts-tls-certificate.yaml`` with the following contents:

   .. code-block:: yaml
      :class: copyable
      :emphasize-lines: 7,12

      # sts-tls-certificate.yaml
      apiVersion: cert-manager.io/v1
      kind: Certificate
      metadata:
        name: sts-certmanager-cert
        namespace: minio-operator
      spec:
        dnsNames:
          - sts
          - sts.minio-operator.svc
          - sts.minio-operator.svc.cluster.local # Replace cluster.local with the value for your domain.
        secretName: sts-tls
        issuerRef:
          name: minio-operator-ca-issuer

   .. important::
   
      The ``spec.secretName`` is not optional.
   
      The secret name **must** be ``sts-tls``.

>       Confirm this by setting ``spec.secretName: sts-tls`` as highlighted above.

Reviewer 3: ```suggestion
      Confirm this by setting ``spec.secretName: sts-tls`` as highlighted in the Certificate YAML.
```
directional text ...
